### PR TITLE
Home task 07 completed

### PR DESCRIPTION
**Pull request by jhonfmg7**
Home task 07 completed,
Project created with "create-react-app"
adding new functionalities: Add new Movie, Edit Movie, and Delete Movie. each functionality connected with backend Movies API.
Captura de pantalla 2022-12-02 a la(s) 10 45 00 a m

### CODE REVIEW

File: src/components/Messages/index.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import * as React from "react";
+import { useSelector } from "react-redux";
+import styles from "../../css-modules/messages.module.css";
+
+// Interface
+import { RootState } from "../../types/redux.interface";
+
+type State = {
+    message: any
```
Comment (mmarkelov) [inline src/components/Messages/index.tsx]: plz avoid `any`

File: src/components/Messages/index.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+function Messages() {
+  // Redux State Extraction
+  const { message } = useSelector<RootState, State>((state) => state.movies);
+
+  if (message) {
```
Comment (mmarkelov) [inline src/components/Messages/index.tsx]: it's better to go with this approach:
```js
return condition ? <section>...</section> : null;
```

File: src/components/main/index.tsx [hunk 3]
```diff
@@ -17,36 +17,34 @@ import Loader from "../loader";
...
+    movies, moviesQuantity, reload, loading, typeSelected, sortBySelected,
+  } = useSelector<RootState, State>((state) => state.movies);
 
   React.useEffect(() => {
-    if (typeSelected === "all") {
-      dispatch(getAllMoviesAction("0", sortBySelected));
-    } else {
-      dispatch(getAllMoviesByGenreAction("0", typeSelected, sortBySelected));
+    if (reload) {
+      if (typeSelected === "all") {
```
Comment (mmarkelov) [inline src/components/main/index.tsx]: please do not use magic strings or numbers. Use constants instead.
The best approach here will be change `CATEGORIES` to enum i guess

File: src/components/main/navbar/index.tsx [hunk 4]
```diff
@@ -33,7 +37,7 @@ function NavBar(props: Props) {
       <div className={styles.sortBySelect}>
         <p className={styles.secondaryText}>Sort by</p>
         <div className={styles.selectContainer}>
-          <select className={styles.selectInput} value={sortBySelected} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setSortBySelected(e.target.value)}>
+          <select className={styles.selectInput} value={sortBySelected} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => dispatch(setSortBySelectedAction(e.target.value))}>
```
Comment (mmarkelov) [inline src/components/main/navbar/index.tsx]: type is redundant here
```suggestion
          <select className={styles.selectInput} value={sortBySelected} onChange={e => dispatch(setSortBySelectedAction(e.target.value))}>
```

File: src/components/modals/AddOrEditMovieForm/Select.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import * as React from "react";
+import { FormikErrors, FormikTouched } from "formik";
+import styles from "../../../css-modules/modal.module.css";
+
+// Interfaces
+import Movie from "../../../types/movie.interface";
+
+interface Props {
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/Select.tsx]: This interface is almost the same as in `Input` component. It's better to create a one interface for this

File: src/components/modals/AddOrEditMovieForm/Select.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      <select
+        id={id}
+        name={value}
+        value={info[value]}
+        className={errors[value] && touched[value] ? styles.inputWithError : styles.input}
+        multiple
+        onChange={handleChange}
+        onBlur={handleBlur}
+      >
+        <option value="Action">Action</option>
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/Select.tsx]: The better approach here will be to create a constant with array of options and then iterate through it.
```js
const OPTIONS = [""Advanture", ..., "Crime"];

...

OPTIONS.map(option => <option key={option} value={option}>{option}</option>)
```

File: src/components/modals/AddOrEditMovieForm/Select.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        <option value="Adventure">Adventure</option>
+        <option value="Science Fiction">Science Fiction</option>
+        <option value="Horror">Horror</option>
+        <option value="Documentary">Documentary</option>
+        <option value="Comedy">Comedy</option>
+        <option value="Crime">Crime</option>
+      </select>
+      <p className={styles.errorInput}>{errors[value] && touched[value] && errors[value]}</p>
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/Select.tsx]:
```suggestion
      <p className={styles.errorInput}>{touched[value] && errors[value]}</p>
```

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 8]
```diff
@@ -1,68 +1,99 @@
...
-    setInfo((oldValue) => ({ ...oldValue, [e.target.name]: e.target.value }));
   };
 
-  const handleReset = () => setInfo(initialState);
+  const validationSchema = Yup.object({
+    title: Yup.string().required("Title is required"),
+    genres: Yup.array().min(1, "Should select at least 1 genre"),
+    runtime: Yup.number().moreThan(0, "Should be more than 0 minutes"),
+    overview: Yup.string().required("Brief description is necessary"),
+    poster_path: Yup.string().required("The URL image is required"),
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]: shouldn't it be
```suggestion
    poster_path: Yup.string().url().required("The URL image is required"),
```

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 9]
```diff
@@ -1,68 +1,99 @@
...
+          <Input isLarge={false} id="rating" title="Rating" info={values} value="vote_average" type="number" placeholder="7.8" errors={errors} touched={touched} handleChange={handleChange} handleBlur={handleBlur} />
+          <Select id="genres" title="Genres" info={values} value="genres" errors={errors} touched={touched} handleChange={handleChange} handleBlur={handleBlur} />
+          <Input isLarge={false} id="runtime" title="Runtime" info={values} value="runtime" type="number" placeholder="minutes" errors={errors} touched={touched} handleChange={handleChange} handleBlur={handleBlur} />
+          <div className={styles.inputGroupExtraLarge}>
+            <label htmlFor="overview" className={styles.label}>Overview</label>
+            <textarea id="overview" cols={30} rows={10} name="overview" placeholder="Movie Description" className={errors.overview && touched.overview ? styles.inputWithError : styles.input} value={values.overview} onChange={handleChange} />
+            <p className={styles.errorInput}>{errors.overview && touched.overview && errors.overview}</p>
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]:
```suggestion
            <p className={styles.errorInput}>{touched.overview && errors.overview}</p>
```

File: src/components/modals/AddOrEditMovieForm/index.tsx [hunk 10]
```diff
@@ -1,68 +1,99 @@
...
+          </div>
+          <div className={styles.textEnd}>
+            <button className={styles.secondaryButton} type="button" onClick={(e) => handleReset(setValues)}>Reset</button>
```
Comment (mmarkelov) [inline src/components/modals/AddOrEditMovieForm/index.tsx]:
```suggestion
            <button className={styles.secondaryButton} type="button" onClick={() => handleReset(setValues)}>Reset</button>
```

File: src/components/modals/DeleteModal.tsx [hunk 11]
```diff
@@ -1,23 +1,42 @@
...
+    dispatch(deleteMovieAction(id));
+    setIsOpen(false);
+    setIsOpenMenu(false);
+  };
+
+  return (
+    <Modal title={title} setIsOpen={setIsOpen} setIsOpenMenu={setIsOpenMenu}>
+      <p className={styles.deleteMessage}>Are you sure you want to delete this movie?</p>
+      <div className={styles.textEnd}>
+        <button className={stylesHeader.secondaryButton} type="button" onClick={() => handleDelete()}>Confirm</button>
```
Comment (mmarkelov) [inline src/components/modals/DeleteModal.tsx]:
```suggestion
        <button className={stylesHeader.secondaryButton} type="button" onClick={handleDelete}>Confirm</button>
```

File: src/components/modals/Modal.tsx [hunk 12]
```diff
@@ -16,7 +16,7 @@ const Modal: React.FunctionComponent<React.PropsWithChildren<Props>> = (props: P
 
   const handleClose = () => {
     setIsOpen(false);
-    setIsOpenMenu(false);
+    setIsOpenMenu && setIsOpenMenu(false);
```
Comment (mmarkelov) [inline src/components/modals/Modal.tsx]:
```suggestion
    setIsOpenMenu?.(false);
```

File: src/redux/actions/moviesAction.ts [hunk 13]
```diff
@@ -1,11 +1,51 @@
...
+const clearMessages = () => ({
+    type: CLEAR_MESSAGES
+});
+
+export function launchReloadAction() {
+    return (dispatch: AppDispatch) => {
+        dispatch(setReload(true));
+        setTimeout(() => {
+            dispatch(clearMessages());
+        }, 3000);
```
Comment (mmarkelov) [inline src/redux/actions/moviesAction.ts]: plz use constant

File: src/redux/actions/moviesAction.ts [hunk 14]
```diff
@@ -55,3 +95,120 @@ const getAllMoviesError = (error: Error) => ({
...
 });
+
+export function createNewMovieAction(info: Movie, setIsOpen: (newState: boolean) => void) {
+    return async(dispatch: AppDispatch) => {
+        dispatch(createMovieStart());
+        
+        try {
+            const response = await fetch(`${ BACKEND_URL }`, {
+                method: "POST",
+                mode: "cors",
```
Comment (mmarkelov) [inline src/redux/actions/moviesAction.ts]: You use this options for all of your requests. You can use it like this:
```js
const COMMON_OPTIONS = {
  mode: "cors",
  cache: "no-cache",
  credentials: "same-origin",
  headers: {
    'Content-Type': 'application/json'
  }
}
```

File: src/redux/reducers/moviesReducer.ts [hunk 15]
```diff
@@ -1,16 +1,25 @@
...
 
 interface InitialState {
     movies: Movie[],
     moviesQuantity: number,
+    reload: boolean,
+    typeSelected: string,
+    sortBySelected: string,
     loading: boolean,
-    error: Error | any
+    error: Error | any,
```
Comment (mmarkelov) [inline src/redux/reducers/moviesReducer.ts]: plz avoid `any`

File: src/components/Messages/index.tsx (hunk 1 above)
Comment (jhonfmg7) [reply src/components/Messages/index.tsx]: resolved

File: src/components/Messages/index.tsx (hunk 2 above)
Comment (jhonfmg7) [reply src/components/Messages/index.tsx]: Ok, I got it!

File: src/components/main/index.tsx (hunk 3 above)
Comment (jhonfmg7) [reply src/components/main/index.tsx]: Resolved

File: src/components/main/navbar/index.tsx (hunk 4 above)
Comment (jhonfmg7) [reply src/components/main/navbar/index.tsx]: Got it

File: src/components/modals/AddOrEditMovieForm/Select.tsx (hunk 5 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/Select.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/Select.tsx (hunk 6 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/Select.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/Select.tsx (hunk 7 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/Select.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 8 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 9 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/AddOrEditMovieForm/index.tsx (hunk 10 above)
Comment (jhonfmg7) [reply src/components/modals/AddOrEditMovieForm/index.tsx]: resolved

File: src/components/modals/DeleteModal.tsx (hunk 11 above)
Comment (jhonfmg7) [reply src/components/modals/DeleteModal.tsx]: resolved

File: src/components/modals/Modal.tsx (hunk 12 above)
Comment (jhonfmg7) [reply src/components/modals/Modal.tsx]: resolved

File: src/redux/reducers/moviesReducer.ts (hunk 15 above)
Comment (jhonfmg7) [reply src/redux/reducers/moviesReducer.ts]: resolved

File: src/redux/actions/moviesAction.ts (hunk 14 above)
Comment (jhonfmg7) [reply src/redux/actions/moviesAction.ts]: Resolved

File: src/components/main/index.tsx [hunk 16]
```diff
@@ -17,36 +17,37 @@ import Loader from "../loader";
...
   moviesQuantity: number,
-  loading: boolean
+  reload: boolean,
+  loading: boolean,
+  typeSelected: string,
+  sortBySelected: string
 }
 
+const NUM_PAGE: string = "0";
+const CATEGORIES: string[] = ["all", "action", "documentary", "comedy", "horror", "crime"];
```
Comment (mmarkelov) [inline src/components/main/index.tsx]: I've got the same constant in `navbar`

File: src/redux/actions/moviesAction.ts (hunk 13 above)
Comment (jhonfmg7) [reply src/redux/actions/moviesAction.ts]: Resolved

File: src/components/main/index.tsx [hunk 17]
```diff
@@ -17,36 +17,37 @@ import Loader from "../loader";
...
   movies: Movie[],
   moviesQuantity: number,
-  loading: boolean
+  reload: boolean,
+  loading: boolean,
+  typeSelected: string,
+  sortBySelected: string
 }
 
+const NUM_PAGE: string = "0";
```
Comment (mmarkelov) [inline src/components/main/index.tsx]: The type definition in this case is redundant , TS is smart enough to understand that it's `string`

File: src/components/main/index.tsx (hunk 16 above)
Comment (jhonfmg7) [reply src/components/main/index.tsx]: Resolved

File: src/components/main/index.tsx (hunk 17 above)
Comment (jhonfmg7) [reply src/components/main/index.tsx]: I got it! thank you

File: src/components/main/index.tsx [hunk 18]
```diff
@@ -17,36 +17,37 @@ import Loader from "../loader";
...
+  } = useSelector<RootState, State>((state) => state.movies);
 
   React.useEffect(() => {
-    if (typeSelected === "all") {
-      dispatch(getAllMoviesAction("0", sortBySelected));
-    } else {
-      dispatch(getAllMoviesByGenreAction("0", typeSelected, sortBySelected));
+    if (reload) {
+      if (typeSelected === CATEGORIES[0]) {
+        dispatch(getAllMoviesAction(NUM_PAGE, sortBySelected));
```
Comment (mmarkelov) [inline src/components/main/index.tsx]: Also as i can see `getAllMoviesAction` and `getAllMoviesByGenreAction` almost the same functions. So the best option here will be to make one function and use `genre` as optional param. Smth like this:
```js
export function getAllMoviesByGenreAction(numPage: string,  sortBy: string, genre?: string) {
    return async(dispatch: AppDispatch) => {
        dispatch(getAllMoviesStart());

        try {
            const searchParams = new URLSearchParams({ limit: MOVIES_PER_PAGE, offset: numPage, sortBy, sortOrder: SORT_ORDER, filter: genre });
            const response = await fetch(`${ BACKEND_URL }?${searchParams}`);
            const data = await response.json();

            if (response.status === 200) {
                dispatch(getAllMoviesSuccess(data.data, data.totalAmount));
            }
        } catch (error) {
            dispatch(getAllMoviesError(error))
        }
    }
}
```

File: src/redux/actions/moviesAction.ts [hunk 19]
```diff
@@ -53,5 +114,98 @@ const getAllMoviesSuccess = (movies: Movie[], quantity: number) => ({
...
+    payload: {error}
+});
+
+export function createNewMovieAction(info: Movie, setIsOpen: (newState: boolean) => void) {
+    return async(dispatch: AppDispatch) => {
+        dispatch(createMovieStart());
+        
+        try {
+            const response = await fetch(`${ BACKEND_URL }`, { ...COMMON_OPTIONS, method: "POST", body: JSON.stringify(info) });
+            console.log(response);
```
Comment (mmarkelov) [inline src/redux/actions/moviesAction.ts]: plz remove console log

File: src/redux/types/index.ts [hunk 20]
```diff
@@ -1,3 +1,22 @@
-export const GET_MOVIES_START = 'GET_MOVIES_START';
-export const GET_MOVIES_SUCCESS = 'GET_MOVIES_SUCCESS';
-export const GET_MOVIES_ERROR = 'GET_MOVIES_ERROR';
\ No newline at end of file
+export const SET_TYPE_SELECTED: string = 'SET_TYPE_SELECTED';
+export const SET_SORTBY_SELECTED: string = 'SET_SORTBY_SELECTED';
+
+export const SET_RELOAD: string = 'SET_RELOAD';
```
Comment (mmarkelov) [inline src/redux/types/index.ts]: Unnecessary type annotations

File: src/components/main/index.tsx (hunk 18 above)
Comment (jhonfmg7) [reply src/components/main/index.tsx]: Resolved, I need to declare two variables for params, cause if I pass to search params filter with "" or null, the API doesn't get a response

File: src/redux/actions/moviesAction.ts (hunk 19 above)
Comment (jhonfmg7) [reply src/redux/actions/moviesAction.ts]: Sorry I forget the console.log

File: src/redux/types/index.ts (hunk 20 above)
Comment (jhonfmg7) [reply src/redux/types/index.ts]: Got it

**Pull request merged**